cmake/DownloadFolly.cmake

```diff
@@ -0,0 +1,39 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+## Download facebook's folly library. 
+## SOURCE_DIR is typically the cmake source directory
+## BINARY_DIR is the build directory, typically 'build'
+
+function(download_folly SOURCE_DIR BINARY_DIR)
+
+	
+	ExternalProject_Add(
+		facebook-folly-proj
+		GIT_REPOSITORY "https://github.com/facebook/folly.git"
+		GIT_TAG "v2020.05.18.00"
```

Comment:
Any reason you bumped up the versions from the master branch? Probably better to use the same versions and upgrade them in a separate commit to limit the scope of changes?

Reply:
I believe the master branch had 2017.09.04, which gave me concern due to its age. This was one of two reasons why I submitted this as a draft ( the other being that not all tests passed ). It allows me to get feedback on this very item so thanks for pointing this out.

1) On one hand I can keep the versions as-is from master ( thus reducing or eliminating the code changes ). It would be a smaller PR, less to review, and if I were the reviewer I'd probably prefer that :) +1482 -382 is a pain to review IMO and the version bumps cause scope creep.

2) On the other hand bumping versions after such an age gap did require some code resolutions, which I think when the are combined in this PR give perspective.

The other consideration for this when I approached this I'm on a much newer toolchain and am building outside of docker, so managing the numerous dependencies of folly/wangle caused compatibility concerns. I haven't fully validated that this will be a problem so going with #1 and splitting up this PR would make total sense to me.

I can extract the code changes required for the version bump very easily, and ultimately this is why I posted it as a draft-- so I can get this feedback early and get opinions on this. 

Would appreciate @joshelser 's take too RE version bumping. 

Look forward to your responses! Thanks!

Reply:
Ya, I was more talking about the scope creep and a reviewing perspective. I'm totally up for bumping the version but I think it'd be nice to do it in a follow up patch _ if it is not too much work for you_. I'll let @joshelser weigh in, he has more context here than I do.

Reply:
Yep. I'll take a look. I updated folly ( which resulted in a lot of other changes ) because of this
`
https://github.com/facebook/folly/blob/v2017.09.04.00/CMakeLists.txt#L24`

My preference was to use CMake if and when possible but it definitely becomes difficult as we aim to control the dependency tree.

Reply:
Agreed.

Reply:
> On the other hand bumping versions after such an age gap did require some code resolutions, which I think when the are combined in this PR give perspective.

From what I remembered the last time I was in this code, I assumed the version bumps were necessitated to avoid making changes for the sake of _not_ updating folly/wangle (et al) versions. Totally OK if this gets "big" just for the sake of getting back to a good position.